UOP1_Project/Assets/Scripts/PlayerSpawnPoint.cs

```diff
@@ -0,0 +1,48 @@
+﻿using UnityEngine;
+
+/// <summary>
+/// Spawns the player at its own location
+/// </summary>
+public class PlayerSpawnPoint : MonoBehaviour
+{
+    [SerializeField]
+    [Tooltip("If True, the player will be spawned on Awake")]
+    bool spawnOnAwake = true;
+    [SerializeField]
+    [Tooltip("The Prefab of the player (Pig)")]
+    GameObject playerPrefab;
+
+    Transform gameplayCamera;
+    InputReader inputReader;
+
+    /// <summary>
+    /// Gather external components. Throws if it can't find them.
+    /// </summary>
+    void Awake()
+    {
+        if (playerPrefab == null)
+            Debug.LogError("No Player to spawn");
+
+        inputReader = FindObjectOfType<InputReader>();
+        if (inputReader == null)
+            Debug.LogError("No Input Reader has been given to assign to the player");
+
+        gameplayCamera = Camera.main.transform;
+        if (gameplayCamera == null)
+            Debug.LogError("No Gameplay Camera has been given to assign to the player");
+
+        if (spawnOnAwake)
+            SpawnPlayer();
+    }
+
+    /// <summary>
+    /// Spawns the player and puts this GO as child of the newly spawned player.
+    /// </summary>
+    public void SpawnPlayer()
+    {
+        var player = Instantiate(playerPrefab, transform.position, transform.rotation);
+        transform.parent = player.transform;
+        var protagonistComponent = player.GetComponent<Protagonist>();
+        protagonistComponent.Initialise(inputReader, gameplayCamera);
```

Comment:
Initiate components inside the player should be a responsibility of the player itself.
Your spawn script should only do one thing: spawn

Reply:
You are not wrong about this. The spawn script should only do the spawn.
But to achieve what you say, the Protagonist script should be changed upfront. Not sure it's in the scope of this task.

Reply:
Yes, that's why I think your script should only spawn a generic prefab in the scene.
The Protagonist script (or any other one inside the player GameObject) should initialize those components for the player. If so, your script will not need to be refactored in the future.

Reply:
let me try and we'll see what the others think